### PR TITLE
Add test case for pulling image from Pulp Registry

### PR DESCRIPTION
**Pull request by rochacbruno**
closes #4460
https://pulp.plan.io/issues/4460

Required PR: pulp/pulpcore#14
Required PR: PulpQE/pulp-smash#1176

### CODE REVIEW

**Comment by rochacbruno** [comment]
@pulp/qe

**Comment by rochacbruno** [comment]
@dkliban

File: pulp_docker/tests/functional/api/test_pull_content.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        This tests targets the following issue:
+
+        * `Pulp #4460 <https://pulp.plan.io/issues/4460>`_
+        """
+        cls.cfg = config.get_config()
+        cls.client = api.Client(cls.cfg, api.json_handler)
+        cls.teardown_cleanups = []
+
+        # Step 1
+        _repo = cls.client.post(REPO_PATH, gen_repo())
```
[the remark this answers is not in the document]
Comment (rochacbruno) [reply pulp_docker/tests/functional/api/test_pull_content.py]: This is a temporary variable, on other tests we use `self.repo` other variable.

File: pulp_docker/tests/functional/api/test_pull_content.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        4. Create a docker publisher.
+        5. Use the publisher to create a publication.
+        6. Create a docker distribution to serve the publication.
+
+        This tests targets the following issue:
+
+        * `Pulp #4460 <https://pulp.plan.io/issues/4460>`_
+        """
+        cls.cfg = config.get_config()
+        cls.client = api.Client(cls.cfg, api.json_handler)
```
[the remark this answers is not in the document]
Comment (rochacbruno) [reply pulp_docker/tests/functional/api/test_pull_content.py]: So this needs a fix on other plugins see https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/tests/functional/api/test_download_content.py#L61

File: pulp_docker/tests/functional/api/test_pull_content.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    DOCKER_UPSTREAM_TAG,
+)
+from pulp_docker.tests.functional.utils import set_up_module as setUpModule  # noqa:F401
+
+
+class PullContentTestCase(unittest.TestCase):
+    """Verify whether images served by Pulp can be pulled."""
+
+    @classmethod
+    def setUpClass(cls):
```
[the remark this answers is not in the document]
Comment (rochacbruno) [reply pulp_docker/tests/functional/api/test_pull_content.py]: I can add a `try` block on the whole code here with a broader exception handling.

The other option would be having `addCleanup` but that would require recreating all objects for each test case.

File: pulp_docker/tests/functional/api/test_pull_content.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kersommoura) [reply pulp_docker/tests/functional/api/test_pull_content.py]: @rochacbruno, you are right. We should use more `page_handler` for Pulp 3.

**Comment by rochacbruno** [comment]
failures due to https://github.com/pulp/pulp_docker/pull/322

File: .travis/install.sh [hunk 4]
```diff
@@ -29,5 +30,35 @@ fi
...
+sudo apt-get update -y
+sudo apt-get install -y podman
+
+# Configure podman for insecure registry
+sudo bash -c 'cat << EOF > /etc/containers/registries.conf
+[registries.search]
+registries = ["docker.io", "registry.fedoraproject.org", "quay.io", "registry.access.redhat.com", "registry.centos.org"]
+
+[registries.insecure]
+registries = ["localhost:8080", "localhost:8000"]
```
Comment (ipanova) [inline .travis/install.sh]: what's is the reason of having localhost on  8000?

File: pulp_docker/tests/functional/api/test_pull_content.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        registry.pull(DOCKER_UPSTREAM_NAME)
+        self.teardown_cleanups.append((registry.rmi, DOCKER_UPSTREAM_NAME))
+        remote_image = registry.inspect(DOCKER_UPSTREAM_NAME)
+
+        self.assertEqual(
+            local_image[0]['Digest'],
+            remote_image[0]['Digest']
+        )
+
+    def test_pull_image_with_tag(self):
```
Comment (ipanova) [inline pulp_docker/tests/functional/api/test_pull_content.py]: i suggest to also add pull image by digest

File: pulp_docker/tests/functional/api/test_pull_content.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+            any(
+                [
+                    result['blobSum'] in checksums
+                    for result in get_docker_hub_remote_blobsums()
+                ]
+            ),
+            'Cannot find a matching layer on remote registry.'
+        )
+
+    def test_pull_image(self):
```
Comment (ipanova) [inline pulp_docker/tests/functional/api/test_pull_content.py]: @rochacbruno maybe this suggestion will be out of the scope if the testcase but please consider to add tests where the fetch of the content goes directly through API calls.
podman and docker are cli clients that already send some headers to the registry and have their own logic; a lot of people intergrate with the registries by directly sending API calls where they define accept headers.
The reason i am suggesting to add these tests is that we are sure we return correct content( if  availability of course) based on incoming headers. For example we would cover this bug https://pulp.plan.io/issues/4314

Comment (ipanova) [reply pulp_docker/tests/functional/api/test_pull_content.py]: this section might be helpful to understand how registry interprets incoming accept_headers and what it decided to return. https://docs.docker.com/registry/spec/manifest-v2-2/#backward-compatibility

Comment (kersommoura) [reply pulp_docker/tests/functional/api/test_pull_content.py]: @rochacbruno, @ipanova, I filed another test case testing what was suggested here.

https://pulp.plan.io/issues/4531

File: .travis/install.sh (hunk 4 above)
Comment (rochacbruno) [reply .travis/install.sh]: well no need.. it was just a copy/paste.. fixing it

File: pulp_docker/tests/functional/api/test_pull_content.py (hunk 6 above)
Comment (rochacbruno) [reply pulp_docker/tests/functional/api/test_pull_content.py]: Thanks @ipanova your suggestions will come in new PRs

File: pulp_docker/tests/functional/api/test_pull_content.py (hunk 5 above)
Comment (rochacbruno) [reply pulp_docker/tests/functional/api/test_pull_content.py]: Will also land in a new PR later

**Comment by rochacbruno** [comment]
Travis is passing on fork, but github is having trouble to update the status here
https://travis-ci.org/rochacbruno/pulp_docker/builds/505292049

**Comment by rochacbruno** [comment]
All passing now

**Comment by kersommoura** [comment]
I had problems with my pulp 3 installation, but I will run those tests locally once more before merge them.

**Comment by ipanova** [comment]
waiting on @kersommoura  ack to merge :)

**Comment by kersommoura** [comment]
@ipanova, please, can you merge this one?

**Comment by dkliban** [comment]
This PR needs to be rebased against master.

**Comment by rochacbruno** [comment]
@dkliban rebased

**Pull request merged**